### PR TITLE
wgsl: support arrayLength on pointer to sized array

### PR DESCRIPTION
**Pull request by dneto0**
Addresses the consensus subset of #2074

Change-Id: I24737e2d3c961cb4f61c61e9d6ee522fdbe025a9

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -6887,6 +6888,13 @@ That's not a full user-defined function declaration.
     <td>Test a normal value according to [[!IEEE-754|IEEE-754]].<br>
     [=Component-wise=] when |T| is a vector.
 
+  <tr algorithm="fixed-sized array length">
+    <td>|e|: ptr&lt;storage,array&lt;|T|,|N|&gt;&gt;
```
Comment (alan-baker) [inline wgsl/index.bs:6892]: Copy-paste bug, should have a parameter for the storage class.

**Review by alan-baker — CHANGES_REQUESTED** [review]

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-09-07-Minutes#wgsl-support-arraylength-on-pointer-to-sized-array-2081-">WGSL meeting minutes 2021-09-07</a></summary>



* DN: Ben followed up here, original poster wasn’t aware of the ability to resize arrays based on global lets (not implemented in Tint?), and their usecase didn’t need it(?), so our purpose here is more tenuous.
* DN: Also, some of us don’t like the pointer form, but also, some of us don’t like the non-pointer form. Particularly having to maybe (officially) do the evaluation of the expression in order to determine the size.
* DN: As such, we’re not sure we need this, so we’d like to push this post-MVP.
* MM: A little sad about this. We’d like to see this eventually, and we’re ready to talk about this in more depth now, but we’re ok with pushing it back.
* DM: Why did people not like the pointer form?
* DN: I’m trying to represent another person’s stance, but people don’t like the extra character/typing to do this.
* MM: Ideally we’d like to see this for all array types. For now, because of the shape of things, it’s easier to not allow it on `let` arrays for now, even if we want to go further eventually.
* JG: I’m having trouble entirely following here, so examples of good-for-me-thing vs bad-for-me-thing, that’d be easier to follow for me.
* DN: I’m trying to capture the warning here about disadvantages to the pointer approach, but not to block it. We will definitely have to keep explaining this, but maybe that’s fine.
* MM: I’ll echo that. Also the language is worse off if theres *no* way to do this. *A* way to do this would be an improvement here.
* DN: Ok, ok to land the pointer form? (yes)
* **Result: Land pointer form, MM to open issues for the other forms.**
* DM: I hope we don’t need more forms, ideally.
* MM: It’s a good point that we should consider all this in respect to the rest of the language.


</details>

**Comment by ben-clayton** [comment]
> DM: Why did people not like the pointer form?
> DN: I’m trying to represent another person’s stance, but people don’t like the extra character/typing to do this.

The reason is more than just the extra typing. Unlike runtime sized arrays, which must exist in a SSBO, and therefore always has memory, fixed-size arrays can exist without storage (`let` declarations).
You cannot take the address of a value without storage, so you can't use `arrayLength()` on a `let` array or on the returned value of a function call. Forcing the array value into a `var` seems completely broken given that this builtin is pure compile-time-eval for fixed-size arrays.

The argument I've heard as to why `arrayLength()` takes a pointer *for runtime sized arrays*, is to do with having the builtin conform to the load rules (you can't load a runtime sized array). I'm not sure how compelling that argument is given that there are numerous builtins that already have exceptions to the standard rules - including `arrayLength` itself, as you cannot normally have a parameter be a pointer to a `storage` storage class.

If we changed the signatures to accept an array value, then users wouldn't need to type the perplexing address-of-expr, and array `let`s and return values would also be accepted. Which is great...

**However**

There's a bag of complexity around whether the argument should be runtime evaluated or not.

I believe the WG leaned towards the stance that arguments should always be evaluated (unlike c++'s `sizeof`). So `arrayLength(foo())` would evaluate to both a compile-time evaluated constant, and something akin to `ignore(foo())`. This involves transforming a single expression into multiple statements, which carries a lot of complexity.
For example, consider:  

```
for(var i = 0; i < arrayLength(foo()); i = i + 1) {
```

Where would `ignore(foo)` get placed?

Having `arrayLength` act more like `sizeof` would fix these issues, but the WG has already expressed dislike for this.

So to summarize my thoughts:
* Having `arrayLength` take a pointer-to-array is plain weird for the user to have to type, and I'm not sold on the load-rule being a good justification.
* Requiring a pointer prevents `arrayLength` being used on array `let`s, and function return values. This is a weird usability constraint.
* Forcing array values into `var`s just so you can take the address seems like a really bad design choice for a pure compile-time eval builtin.
* If we added / changed `arrayLength` to accept non-pointers, there's a bunch of complexities around evaluation of the argument.
* I'm opposed to having two ways to do the same thing (both pointer and non-pointer overloads)
* I think landing this right now is rushing into a language feature we'd regret and likely end up removing. I don't think there is any real need this right now.
* I'd prefer we held of submitting this until we've properly considered the above.

**Comment by dneto0** [comment]
Rebased and fixed @alan-baker's feedback.

File: wgsl/index.bs (hunk 1 above)
Comment (dneto0) [reply wgsl/index.bs]: Thanks. Fixed now.

**Comment by kvark** [comment]
@ben-clayton does appear to have strong views here, and my reading is that the suggestion is to postpone this. Especially since users could use non-overridable module-scope constants for array sizes (something I missed in yesterdays discussion, and @ben-clayton pointed to on matrix).
  - So, in all cases where the size of an array is known at compile time, they can just declare a constant instead of taking `arraySize`
  - Similarly, if it's overridable by a pipeline constant, taking the size isn't necessary either, since there is already a constant to use
  - This leaves the only problematic case where the size is setup by any other mechanism than a constant (depending on #1431 resolution)

With that in mind, we should indeed postpone adding this until #1431 is closer to resolution.

**Comment by dneto0** [comment]
Closing.  Let's revisit the discussion on the original issue.

**Pull request closed without merging**